### ISSUE TITLE
subuid and subgid check?

### ISSUE THREAD

**Issue by ikke-t** [linux-system-roles/podman#29]
I quicly read through the tasks, and this caught my eye: If /etc/sub[gu]id files don't list the user, podman will fail. I had the trouble in my role. Not all the tools add user to the files, like anaconda and if authorized by freeipa/IdM.

It might make sense to somehow at least check and notify about it. Better yet, add the line for the user.

Here's what I mean: https://github.com/ikke-t/podman-container-systemd/blob/master/tasks/check_subid.yml

**Comment by richm** [comment]
ok - we'll have to do this in a loop for every user mentioned in any of the list of kube specs passed in
also - what to do about `    line: "{{ container_run_as_user }}:165536:65536"` - do we need a different, unique number for each unique user instead of hardcoding `165536:65536`?

**Comment by ikke-t** [comment]
it should be different. Otherwise they all end up using the same space. I don't know how useradd does it. Is there a mechanism to ask useradd or any other command to generate the line?

**Comment by ikke-t** [comment]
The use cases that I've run into this recently is that when I set up our lab to use IdM (freeipa) with central auth management, the system doesn't create the sub stuff when you just login to host. I don't know if there was pam module to do the entries in such case.

Also in RH edge boxes do miss that. That's likely quite typical to run podman in edge boxes. One of the likely places to get the request from one day. The user creation there is likely done by anaconda or image builder customization. They use /usr/lib/passw instead od /erc/passwd, so user creation is a bit different anyhow.

**Comment by giuseppe** [comment]
if there are no additional IDs available, Podman can only use one single ID so you'd have to set to ignore chown errors in the `storage.conf` file.  Said so that is not optimal configuration since all IDs in the container are squashed to the same one.

freeipa supports remote assignment of additional IDs to the users and podman/new?idmap would pick them through libsubid.  Could that be used?

**Comment by richm** [comment]
> if there are no additional IDs available, Podman can only use one single ID so you'd have to set to ignore chown errors in the `storage.conf` file. Said so that is not optimal configuration since all IDs in the container are squashed to the same one.
> 
> freeipa supports remote assignment of additional IDs to the users and podman/new?idmap would pick them through libsubid. Could that be used?

If that isn't a suitable solution - what can/should we do in the podman role?  Should we implement something like https://github.com/ikke-t/podman-container-systemd/blob/master/tasks/check_subid.yml?

**Comment by vrothberg** [comment]
@richm, a check like that would improve the user experience.  I think it's worth doing it and guide the user to resolve the issue.  It's only required for rootless users.

**Comment by richm** [comment]
> @richm, a check like that would improve the user experience. I think it's worth doing it and guide the user to resolve the issue. It's only required for rootless users.

ok - let me see if I understand using pseudo code
```
for user in rootless_users:
  if user not in /etc/subuid:
    add user to /etc/subuid - $user:$sub_uid_num:$sub_uid_count
for grp in rootless_groups:
  if grp not in /etc/subgid:
    add grp to /etc/subgid - $grp:$sub_gid_num:$sub_gid_count
```
my problem is - how to I get the values to use for `$sub_uid_*` and `$sub_gid_*`?

**Comment by vrothberg** [comment]
You can use `podman info --format` for that:
```
~ $ podman info --format "{{.Host.IDMappings}}"
{[{0 1000 1} {1 100000 65536}] [{0 1000 1} {1 100000 65536}]}
```
If the output shows empty lists, I'd throw an error/warning.

**Comment by briansmith0** [comment]
From what I understand based on the RHEL 8.6 release notes, "Podman now has ability to source subUID and subGID ranges from IdM".  So for a IdM connected system, it seems like we would not want to have the role add subUID/GID information locally. 

useradd on RHEL 8.6 automatically adds lines to the subUID/GID file

The "ansible.builtin.user" module also appears to automatically add lines to the subUID/GID file (just tried this on RHEL 8.6 with Ansible Core 2.12. )

It also appears that the user I created during a RHEL 8.6 interactive install with Anaconda was automatically added to the subUID/GID file (I would need to do another RHEL install to make 100% sure of this)

I tried useradd on a RHEL 7.9 system, and it also adds the entries to subUID/GID, so presumably systems originally built on RHEL 7 and upgraded to RHEL 8/9 would also have subUID/GID setup.  

What circumstances to we anticipate that a user exists on the system but doesn't have the subUID/GID entries setup?
